### PR TITLE
AVRO-3340: Add codeanalysis for all C# projects

### PR DESCRIPTION
**Pull request by zcsizmadia**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3340) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AVRO-3340
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by zcsizmadia** [comment]
Current state warnings and info detected by intellisense:
image

**Comment by zcsizmadia** [comment]
The idea is to fix those warnings and info messages in separate tickets and PRs. Fix might mean modifying th actual source code (preffered) or setting the level to none for a message which can be ignored.

File: .editorconfig
Comment (KyleSchoonover) [inline .editorconfig:78]: csharp_style_var_for_built_in_types = false:error

We should be making it strongly typed.

File: .editorconfig
Comment (KyleSchoonover) [inline .editorconfig:79]: `csharp_style_var_when_type_is_apparent = false:error`
We should be making it strongly typed.

File: .editorconfig
Comment (KyleSchoonover) [inline .editorconfig:80]: `csharp_style_var_elsewhere = false:error`
We should be making it strongly typed.

File: .editorconfig
Comment (KyleSchoonover) [inline .editorconfig:99]: There is no reason to prefix a static field.

File: .editorconfig
Comment (KyleSchoonover) [inline .editorconfig:108]: We should decide whether or not to enforce this., base., etc or use this underscore as the prefix.

**Review by KyleSchoonover — COMMENTED** [review]
Going to need to review more.  I 100% agree the project needs this, but we need to be specific as possible, and be more severe about enforcing these rules.  This is partially, why I am doing some code clean up right now in order to enforce these rules going forward.

**Comment by zcsizmadia** [comment]
@KyleSchoonover As I mentioned in the ticket this is the same style as the MS dotnet team is using. I would prefer  to follow them as much as possible, however allowing certain overrides. Keep in mind some of these overrides atre personal preferences as well.

1. I added an override section to the codeanalysis files
2. Please add your suggestions to the rules to the ticket, it would be easier to track them and make the changes after further  discussions on those issues
3

**Comment by KyleSchoonover** [comment]
@zcsizmadia I too prefer a massive cleanup.  But in order to make things move forward I'm already breaking things out into manageable pull requests.  If we do an all at once it will sit there and probably not get merged.  I am currently going through Avro.main.

**Comment by zcsizmadia** [comment]
> @zcsizmadia I too prefer a massive cleanup. But in order to make things move forward I'm already breaking things out into manageable pull requests. If we do an all at once it will sit there and probably not get merged. I am currently going through Avro.main.

It might make sense to do any code style rework after the code style guidelines are set by using set (by using codestyle files), otherwise different coding styles might  be used , especially if the rework takes too long. Honestly I added this PR because I saw that you started to do individual changes, and I had the concerns I jsst mentioned.

**Comment by zcsizmadia** [comment]
This PR should not be rushed for sure. And I support to override some of the coding syles, to divert from the MS way. however it is IMO smart thing to do flollwing them. If you look at the Coding style src file, I already modded 'IDE0008' to none (Use explicit type). My point was that the baseline should be the MS dotnet team standard + overrides to divert from that.

**Review by KyleSchoonover — COMMENTED** [review]
Updated review comments

**Comment by KyleSchoonover** [comment]
One more quick discussion.  This is only usable in VS2019, and I believe Rider uses it as well.  I believe the latest build tools will take this into account as well.   I just want to call this out, that these are not used in older versions of Visual Studio.

**Comment by zcsizmadia** [comment]
Rider definetley supports it, Visual Studio Code as well. Visual Studio, I am not sure what is the oldest version with Intellisense support

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you, @zcsizmadia !

**Pull request merged**
